Fix CI unit tests

DanielEbert opened the pull request:
In `test_loads_binary_without_crashing()` and `test_loads_binary_with_core_without_crashing()` the GDB output changed. This PR updates the `expected` GDB output in `test_loads.py`.

Currently the CI tests use a 3 year old `bash` binary + core dump and we get a bunch or warnings (see below) if we open this core dump in GDB. I assume GDB or the CI environment does not support this `bash` binary and/or crash dump.
For this reason, this PR replaces the `bash` binary + crash dump with: `tests/binaries/div_zero_binary/binary.c` and a crash dump based on this div_zero_binary. `test_loads.py` is updated accordingly.

Warnings if we open the `bash` binary with its core dump in GDB:
~~~python
[
    'Reading symbols from %s...' % BASH_BIN,
    '(No debugging symbols found in %s)' % BASH_BIN,
    '',
    "warning: Can't open file /home/user/pwndbg/tests/corefiles/bash/binary "
    'during file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libnss_files-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libnss_nis-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libnsl-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libnss_compat-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libc-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/libdl-2.19.so during "
    'file-backed mapping note processing', '',
    "warning: Can't open file /lib/x86_64-linux-gnu/ld-2.19.so during file-backed "
    'mapping note processing',
    "[New LWP 13562]", '',
    "warning: Unexpected size of section `.reg-xstate/13562' in core file.", '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libtinfo.so.5" is not '
    'at the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libdl.so.2" is not at '
    'the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libc.so.6" is not at '
    'the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib64/ld-linux-x86-64.so.2" is not at the '
    'expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libnss_compat.so.2" is '
    'not at the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libnsl.so.1" is not at '
    'the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libnss_nis.so.2" is not '
    'at the expected address (wrong library or version mismatch?)', '',
    'warning: .dynamic section for "/lib/x86_64-linux-gnu/libnss_files.so.2" is '
    'not at the expected address (wrong library or version mismatch?)',
    "Core was generated by `/home/user/pwndbg/tests/corefiles/bash/binary'.",
    'Program terminated with signal SIGINT, Interrupt.', '',
    "warning: Unexpected size of section `.reg-xstate/13562' in core file.",
    '#0  0x00007ffff76d36b0 in ?? ()',
]
~~~

DanielEbert committed:
fix CI tests

DanielEbert committed:
replace "bash" binary with "div_zero_binary" in CI tests

disconnect3d commented:
Hey, thanks a lot for this PR, but unfortunately it can't be merged in its current form as we can't be sure if the binaries are not malicious and people may run them locally on their machines in the future. I hope you understand why we can't trust anyone in here.

We probably have to build those binaries in a docker container, or, run gcc before tests if possible? But IIRC this may mean we will change the binaries and addresses and wont be able to succeed with tests - is that right?

DanielEbert committed:
compile binary just-in-time

DanielEbert commented:
Hi, that is a good point.

You are right, in the `output` the following can change:
- '45429' in `'[New LWP 45429]'`
- '0000556e99b7913c' in `'#0 0x0000556e99b7913c in main ()'`

We could use regular expressions to check these lines.

In the 'compile binary just-in-time' commit, I added these regex checks, removed the binary, and run gcc before the tests.

disconnect3d commented:
Nice! Hmm, does it then make sense to keep the `core` file in the repo too? Can we maybe generate it as well? (otherwise, it will always be out-of-sync and kinda may not make sense to load it at all?)

DanielEbert committed:
generate core just-in-time

DanielEbert committed:
generate core just-in-time

DanielEbert committed:
remove .swp file

DanielEbert commented:
That is probably a good call. We can generate the core file for example with GDB. We can run the binary in GDB, wait until the fault is triggered, and use GDB's generate-core-file command to dump the memory to a file (see e4deab5).

If I do that, GDB outputs a warning (GDB output below). I think GDB tries and fails to read 4k bytes of the vsyscall mapping. The vsyscall only maps 1k bytes so I'm not too sure what that is about. Anyway, it generates a core file and if we open this core file in GDB (as far as I can tell) it looks good.

GDB output:
~~~
$ /tmp/div# gdb binary
... copyright message
Reading symbols from binary...
(No debugging symbols found in binary)
(gdb) run
Starting program: /tmp/div/binary 
Program received signal SIGFPE, Arithmetic exception.
0x000056445d33313c in main ()
(gdb) info proc mappings
process 3731
Mapped address spaces:

          Start Addr           End Addr       Size     Offset objfile
      0x56445d332000     0x56445d333000     0x1000        0x0 /tmp/div/binary
      0x56445d333000     0x56445d334000     0x1000     0x1000 /tmp/div/binary
      0x56445d334000     0x56445d335000     0x1000     0x2000 /tmp/div/binary
      0x56445d335000     0x56445d336000     0x1000     0x2000 /tmp/div/binary
      0x56445d336000     0x56445d337000     0x1000     0x3000 /tmp/div/binary
      0x7fc6584f2000     0x7fc6584f4000     0x2000        0x0 
      0x7fc6584f4000     0x7fc65851a000    0x26000        0x0 /usr/lib/x86_64-linux-gnu/libc-2.33.so
      0x7fc65851a000     0x7fc658685000   0x16b000    0x26000 /usr/lib/x86_64-linux-gnu/libc-2.33.so
      0x7fc658685000     0x7fc6586d1000    0x4c000   0x191000 /usr/lib/x86_64-linux-gnu/libc-2.33.so
      0x7fc6586d1000     0x7fc6586d4000     0x3000   0x1dc000 /usr/lib/x86_64-linux-gnu/libc-2.33.so
      0x7fc6586d4000     0x7fc6586d7000     0x3000   0x1df000 /usr/lib/x86_64-linux-gnu/libc-2.33.so
      0x7fc6586d7000     0x7fc6586e2000     0xb000        0x0 
      0x7fc6586e5000     0x7fc6586e6000     0x1000        0x0 /usr/lib/x86_64-linux-gnu/ld-2.33.so
      0x7fc6586e6000     0x7fc65870d000    0x27000     0x1000 /usr/lib/x86_64-linux-gnu/ld-2.33.so
      0x7fc65870d000     0x7fc658717000     0xa000    0x28000 /usr/lib/x86_64-linux-gnu/ld-2.33.so
      0x7fc658717000     0x7fc658719000     0x2000    0x31000 /usr/lib/x86_64-linux-gnu/ld-2.33.so
      0x7fc658719000     0x7fc65871b000     0x2000    0x33000 /usr/lib/x86_64-linux-gnu/ld-2.33.so
      0x7ffdd8804000     0x7ffdd8825000    0x21000        0x0 [stack]
      0x7ffdd8950000     0x7ffdd8954000     0x4000        0x0 [vvar]
      0x7ffdd8954000     0x7ffdd8956000     0x2000        0x0 [vdso]
  0xffffffffff600000 0xffffffffff601000     0x1000        0x0 [vsyscall]
(gdb) generate-core-file core
warning: Memory read failed for corefile section, 4096 bytes at 0xffffffffff600000.
Saved corefile core
~~~

disconnect3d commented:
I think we also need to remove the `tests/binaries/div_zero_binary/core ` from the PR?

DanielEbert commented:
That is correct. I removed the `core` file in the c5e4fa6 commit. I messed up my commits: there are two commits c5e4fa6 e4deab5 that are both called 'generate core just-in-time'. Sorry for the confusion :)

disconnect3d commented on tests/test_loads.py:
```diff
@@ -68,46 +75,78 @@ def test_loads_pure_gdb_without_crashing():
+    assert any([crash_address_line.match(line) for line in output])
+
 
 @pytest.mark.skipif(launched_locally, reason='This test uses binaries compiled on travis builds.')
```
Now that we compile binaries ourself, I guess we can remove this comment too! :)

DanielEbert replied:
I included these comments on purpose but I think you are right, these regexes are easy to understand so I'll remove these comments

DanielEbert committed:
remove "matches for example ..." comments

disconnect3d commented:
Thanks a lot for this!

DanielEbert commented:
Thank you for your feedback :)
